contracts/sol6/reserves/KyberFprReserveV2.sol

```diff
@@ -0,0 +1,349 @@
+            return 0; // pair is not listed
+        }
+
+        uint256 rate;
+        try conversionRatesContract.getRate(token, blockNumber, isBuy, srcQty) returns(uint256 r) {
+            rate = r;
+        } catch {
+            return 0;
+        }
+        uint256 destQty = getDestQty(src, dest, srcQty, rate);
+
+        if (getBalance(dest) < destQty) return 0;
+
+        if (sanityRatesContract != IKyberSanity(0)) {
+            uint256 sanityRate = sanityRatesContract.getSanityRate(src, dest);
+            if (rate > sanityRate) return 0;
+        }
+
+        return rate;
+    }
+
+    function isAddressApprovedForWithdrawal(IERC20 token, address addr)
+        external
+        view
+        returns (bool)
+    {
+        return approvedWithdrawAddresses[keccak256(abi.encodePacked(address(token), addr))];
+    }
+
+    /// @dev return available balance of a token that reserve can use
+    ///      if using weth, call getBalance(eth) will return weth balance
+    ///      if using wallet for token, will return min of balance and allowance
+    /// @param token token to get available balance that reserve can use
+    function getBalance(IERC20 token) public view returns (uint256) {
+        address wallet = getTokenWallet(token);
+        IERC20 usingToken;
+
+        if (token == ETH_TOKEN_ADDRESS) {
+            if (wallet == address(this)) {
+                // reserve should be using eth instead of weth
+                return address(this).balance;
+            }
+            // reserve is using weth instead of eth
+            usingToken = weth;
+        } else {
+            if (wallet == address(this)) {
+                // not set token wallet or reserve is the token wallet, no need to check allowance
+                return token.balanceOf(address(this));
+            }
+            usingToken = token;
+        }
+
+        uint256 balanceOfWallet = usingToken.balanceOf(wallet);
+        uint256 allowanceOfWallet = usingToken.allowance(wallet, address(this));
+
+        return minOf(balanceOfWallet, allowanceOfWallet);
+    }
+
+    function getDestQty(
+        IERC20 src,
+        IERC20 dest,
+        uint256 srcQty,
+        uint256 rate
+    ) public view returns (uint256) {
+        uint256 dstDecimals = getDecimals(dest);
+        uint256 srcDecimals = getDecimals(src);
+
+        return calcDstQty(srcQty, srcDecimals, dstDecimals, rate);
+    }
+
+    /// @dev do a trade
+    /// @param srcToken Src token
+    /// @param srcAmount Amount of src token
+    /// @param destToken Destination token
+    /// @param destAddress Destination address to send tokens to
+    /// @param validate If true, additional validations are applicable
+    function doTrade(
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool validate
+    ) internal {
+        // can skip validation if done at kyber network level
```

Comment:
lets check this assumption
we are moving toward permission less reserves.
maybe this isn't valid anymore.

on the other hand.
the checks here are not really interesting.
and we know Kyber network will not send wrong value.
and if it does we will probably fix it.


@tranvictor can u comment here please?

Reply:
Lets not rely on any trust assumption from the network. Reserves should validate everything before settlement.

Reply:
OK

but for rate there is trust relation
we keep that.
right?

since re calculate rate is high on gas.

@tranvictor

Reply:
No, rate is the most important thing the reserves have to validate otherwise other reserves can trigger a change in the reserve's inventory or some sort in the middle of the process. I think gas cost for rate calculation is not that high.

Reply:
OK
lets validate and see gas cost.

What does it mean "can trigger a change in the reserve's inventory or some sort in the middle of the process" - I don't follow this part

Reply:
![image](https://user-images.githubusercontent.com/12406924/91422316-99663f80-e889-11ea-821a-39b92aac2eb9.png)

This will be an additional 36-37k gas for FPRs, and 7k-84k for APRs. It's especially bad for t2e rate queries, and token decimals aren't 18.

Reply:
![image](https://user-images.githubusercontent.com/12406924/91423373-0928fa00-e88b-11ea-9f24-4959db5f8123.png)

This will be an additional 36-37k gas for FPRs, and 7k-90k for APRs. It's especially bad for t2e rate queries, and token decimals aren't 18.

Reply:
how many steps does each source quantity have for fpr in this table?

Reply:
I've added rate validation in trade func, some gas data from tests (average 15 transactions):

With rate validation in trade function:
1. Eth + Tokens in reserve: 114,000
2. Eth in Reserve, Token in wallet: 122,000
3. Weth: 147,000 (increase around 33k due to transfer + wrap/unwrap)

Without rate validation in trade function:
1. Eth + Tokens in reserve: 75,300
2. Eth in Reserve, Token in wallet: 82,000
3. Weth: 106,000 (increase around 31k due to transfer + wrap/unwrap)

Summary:
1. Using weth costs around 30k-35k more gas for each trade due to transfer + wrap/unwrap
2. Rate validation again in trade function costs 40k more, (on average, depends on number of steps, one step could cost around 2k-3k more according to data from conversion rate contract)

Reply:
OK

Weth costs pretty much.
rate validation around 40k matches in large results from Desmond above.
but seems one step is less then 2-3k

just need to see how much it would be with max steps.
which is 10 qty 10 imbalance for conversionRate
and 16 steps for enhanced conversionRate

Reply:
With/without rate validation:
- No sanity rate:
    - Different for 0 step: 26.7k
    - Gas consumption per step: 2k
    - Min: 26.7 (0 step), Max: 26.7 + 2 * 20 = 66.7k (20 steps)
- Has sanity rate:
    - Different for 0 steps: 31.8k
        - 26.7k for calling conversion rate 0 step
        - 5.1k extra for sanity rate call
    - Gas consumption per step: 2k
    - Min: 31.8k (0 step), Max: 31.8 + 2 * 20 = 71.8k (20 steps)

Reply:
OK
we continue with the rate validation
later we should find a way to save gas for this validation